restapi/stories.go

```diff
@@ -27,35 +27,165 @@ import (
+		names[i] = t.Name
+	}
+	TaggedStory := TaggedStory{story, names}
+	return TaggedStory
+}
+
 func (api api) ReturnAllStories(w http.ResponseWriter, r *http.Request) render.Renderer {
 	var stories []models.Story
 
-	err := api.database.Where("is_visible = true").Find(&stories).Error
+	err := api.database.Preload("Author").Where("is_visible = true").Find(&stories).Error
 	if err != nil {
 		return rest.ErrInternal(api.logger, err)
 	}
 
-	return rest.JSONStatusOK(stories)
+	return rest.JSONStatusOK(api.ReturnTaggedStories(stories))
 }
 
 func (api api) ReturnStoryByID(w http.ResponseWriter, r *http.Request) render.Renderer {
 	var story models.Story
 	id := chi.URLParam(r, "storyID")
 
-	err := api.database.Where("is_visible = true").Where("id=?", id).First(&story).Error
+	err := api.database.Preload("Author").Where("id=?", id).First(&story).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
+			return rest.ErrNotFound(fmt.Sprintf("Could not find story with ID %s", id))
+		}
+		return rest.ErrInternal(api.logger, err)
+	}
+	return rest.JSONStatusOK(api.ReturnTaggedStory(story))
+}
+
+func (api api) EditStoryByID(w http.ResponseWriter, r *http.Request) render.Renderer {
+	var story models.Story
+	id := chi.URLParam(r, "storyID")
+
+	err := api.database.Preload("Author").Where("id=?", id).First(&story).Error
 	if err != nil {
 		if err == gorm.ErrRecordNotFound {
 			return rest.ErrNotFound(fmt.Sprintf("Could not find story with ID %s", id))
 		}
 		return rest.ErrInternal(api.logger, err)
 	}
 
-	return rest.JSONStatusOK(story)
+	imageKey, err := convertImageURLToKeyName(story.ImageURL)
+
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, "ImageUrl of existing image is not valid", err)
+	}
+
+	msg, err := api.DeleteImageInS3(imageKey) //delete existing image
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, msg, err)
+	}
+
+	file, h, err := r.FormFile("image")
+
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, "Error getting the image.", err)
+	}
+	imageURL, err := api.uploadImageTos3(file, h.Size, h.Filename)
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, "Error uploading the image to s3.", err)
+	}
+	defer file.Close()
+
+	var author models.Author
+	author.FirstName = r.FormValue("author_first_name")
+	author.LastName = r.FormValue("author_last_name")
+	author.OriginCountry = r.FormValue("author_country")
+
+	errAuthor := api.database.First(&author).Error
+	if errAuthor != nil {
+		if errAuthor == gorm.ErrRecordNotFound {
+			country := countries.ByName(author.OriginCountry)
+			if country == countries.Unknown {
+				return rest.ErrInvalidRequest(api.logger, "Unknown origin country", nil)
+			}
+			if err := api.database.Create(&author).Error; err != nil { //Adding author
+				return rest.ErrInternal(api.logger, errAuthor)
+			}
+		} else {
+			return rest.ErrInternal(api.logger, errAuthor)
+		}
+	}
+	author.Bio = r.FormValue("bio")
+	api.database.Save(&author)
```

Comment:
Can we wrap all the database operations (save, delete) in a single transaction -> https://gorm.io/docs/transactions.html.

Let's also do it at the end, I don't think we should be mixing db ops + setters for the story